server/events/event_parser.go

```diff
@@ -671,3 +697,145 @@ func (e *EventParser) ParseBitbucketServerPullEvent(body []byte) (pull models.Pu
 	pull, baseRepo, headRepo, user, err = e.parseCommonBitbucketServerEventData(event.CommonEventData)
 	return
 }
+
+// ParseAzureDevopsPullEvent parses Azure Devops pull request events.
+// See EventParsing for return value docs.
+func (e *EventParser) ParseAzureDevopsPullEvent(event azuredevops.Event) (pull models.PullRequest, pullEventType models.PullRequestEventType, baseRepo models.Repo, headRepo models.Repo, user models.User, err error) {
+	pullResource, ok := event.Resource.(*azuredevops.GitPullRequest)
+	if !ok {
+		errMsg := fmt.Sprintf("failed to type assert event.Resource.")
+		err = errors.New(errMsg)
+		return pull, pullEventType, baseRepo, headRepo, user, err
+	}
+	pull, baseRepo, headRepo, err = e.ParseAzureDevopsPull(pullResource)
+	if err != nil {
+		return pull, pullEventType, baseRepo, headRepo, user, err
+	}
+	createdBy := pullResource.GetCreatedBy()
+	if createdBy == nil {
+		err = errors.New("CreatedBy is null")
+		return pull, pullEventType, baseRepo, headRepo, user, err
+	}
+	senderUsername := createdBy.GetUniqueName()
+	if senderUsername == "" {
+		err = errors.New("CreatedBy.UniqueName is null")
+		return pull, pullEventType, baseRepo, headRepo, user, err
+	}
+	switch event.EventType {
+	case "git.pullrequest.created":
+		pullEventType = models.OpenedPullEvent
+	case "git.pullrequest.updated":
+		pullEventType = models.UpdatedPullEvent
+		if pull.State == models.ClosedPullState {
+			pullEventType = models.ClosedPullEvent
+		}
+	default:
+		pullEventType = models.OtherPullEvent
+	}
+	user = models.User{Username: senderUsername}
+	return pull, pullEventType, baseRepo, headRepo, user, err
+}
+
+// ParseAzureDevopsPull parses the response from the Azure Devops API endpoint (not
+// from a webhook) that returns a pull request.
+// See EventParsing for return value docs.
+func (e *EventParser) ParseAzureDevopsPull(pull *azuredevops.GitPullRequest) (pullModel models.PullRequest, baseRepo models.Repo, headRepo models.Repo, err error) {
+	commit := pull.LastMergeSourceCommit.GetCommitID()
+	if commit == "" {
+		err = errors.New("lastMergeSourceCommit.commitID is null")
+		return pullModel, baseRepo, headRepo, err
+	}
+	url := pull.GetURL()
```

Comment:
We need to iterate through the links and pull out the web url. THis URL is the json url. CAn be done later though 

```
{
  "resource": {
    "comment": {
      "_links": {
        "pullRequests": {
          "href": "https://dev.azure.com/lkysow/_apis/git/pullRequests/1" <<<< need this
        }
      }
    },
```